diff --git a/command/markad-standalone.cpp b/command/markad-standalone.cpp
index 4ad82ec..729e209 100644
--- a/command/markad-standalone.cpp
+++ b/command/markad-standalone.cpp
@@ -3654,8 +3654,10 @@ bool cMarkAdStandalone::SaveInfo() {
         if (line[0] == 'X') {
             int stream = 0;
             unsigned int type = 0;
-            char descr[256] = "";
+            char descr[256];
 
+	    memset(descr, 0, sizeof(descr));
+            memset(lang, 0, sizeof(lang));
             int result=sscanf(line, "%*c %3i %3X %3c %250c", &stream, &type, (char *) &lang, (char *) &descr);
             if ((result != 0) && (result != EOF)) {
                 switch (stream) {
@@ -3801,9 +3803,9 @@ bool cMarkAdStandalone::IsVPSTimer() {
     FREE(strlen(fpath)+1, "fpath");
     free(fpath);
 
-    size_t size;
+    size_t size = 0;
     char   *line        = NULL;
-    char   vpsTimer[12] = "";
+    char   vpsTimer[13] = "";
 
     while (getline(&line, &size, mf) != -1) {
         sscanf(line, "%12s", (char *) &vpsTimer);
@@ -3876,9 +3878,13 @@ time_t cMarkAdStandalone::GetRecordingStart(time_t start, int fd) {
             time_t now = time(NULL);
             struct tm tm_r;
             struct tm t = *localtime_r(&now, &tm_r); // init timezone
-            if (sscanf(timestr, "%4d-%02d-%02d.%02d%*c%02d", &t.tm_year, &t.tm_mon, &t.tm_mday, &t.tm_hour, & t.tm_min)==5) {
-                t.tm_year -= 1900;
-                t.tm_mon--;
+	    int y, mo, d, h, mi;
+            if (sscanf(timestr, "%4d-%02d-%02d.%02d%*c%02d", &y, &mo, &d, &h, &mi)==5) {
+		t.tm_year = y - 1900;
+		t.tm_mon = mo - 1;
+		t.tm_mday = d;
+		t.tm_hour = h;
+		t.tm_min = mi;
                 t.tm_sec = 0;
                 t.tm_isdst = -1;
                 isyslog("getting recording start from directory (can be wrong!)");
@@ -3984,10 +3990,11 @@ bool cMarkAdStandalone::LoadInfo() {
     }
 
     char *line = NULL;
-    size_t linelen;
+    size_t linelen = 0;
     while (getline(&line, &linelen, f) != -1) {
         if (line[0] == 'C') {
-            char channelname[256] = "";
+            char channelname[256];
+	    memset(channelname, 0, sizeof(channelname));
             int result = sscanf(line, "%*c %*80s %250c", (char *) &channelname);
             if (result == 1) {
                 macontext.Info.ChannelName = strdup(channelname);
@@ -4024,7 +4031,9 @@ bool cMarkAdStandalone::LoadInfo() {
             }
         }
         if ((line[0] == 'E') && (!bLiveRecording)) {
-            int result = sscanf(line,"%*c %*10i %20li %6i %*2x %*2x", &startTime, &length);
+            long st;
+            int result = sscanf(line,"%*c %*10i %20li %6i %*2x %*2x", &st, &length);
+	    startTime=(time_t)st;
             if (result != 2) {
                 dsyslog("cMarkAdStandalone::LoadInfo(): vdr info file not valid, could not read start time and length");
                 startTime = 0;
@@ -4032,6 +4041,7 @@ bool cMarkAdStandalone::LoadInfo() {
             }
         }
         if (line[0] == 'T') {
+	    memset(title, 0, sizeof(title));
             int result = sscanf(line, "%*c %79c", title);
             if ((result == 0) || (result == EOF)) {
                 title[0] = 0;
@@ -4055,7 +4065,8 @@ bool cMarkAdStandalone::LoadInfo() {
         }
         if ((line[0] == 'X') && (!bLiveRecording)) {
             int stream = 0, type = 0;
-            char descr[256] = "";
+            char descr[256];
+	    memset(descr, 0, sizeof(descr));
             int result=sscanf(line, "%*c %3i %3i %250c", &stream, &type, (char *) &descr);
             if ((result != 0) && (result != EOF)) {
                 if ((stream == 1) || (stream == 5)) {
diff --git a/command/marks.cpp b/command/marks.cpp
index 7f385ee..b949fd7 100644
--- a/command/marks.cpp
+++ b/command/marks.cpp
@@ -559,9 +559,9 @@ int cMarks::LoadVPS(const char *directory, const char *type) {
     free(fpath);
 
     char *line = NULL;
-    size_t length;
-    char typeVPS[15] = "";
-    char timeVPS[20] = "";
+    size_t length = 0;
+    char typeVPS[16] = "";
+    char timeVPS[21] = "";
     int offsetVPS = 0;
     while (getline(&line, &length,mf) != -1) {
         sscanf(line, "%15s %20s %d", (char *) &typeVPS, (char *)&timeVPS, &offsetVPS);
diff --git a/plugin/markad.cpp b/plugin/markad.cpp
index 9973d11..7b96d5b 100644
--- a/plugin/markad.cpp
+++ b/plugin/markad.cpp
@@ -339,7 +339,7 @@ bool cPluginMarkAd::ReadTitle(const char *Directory) {
     }
 
     char *line = NULL;
-    size_t length;
+    size_t length = 0;
     while (getline(&line, &length, f) != -1) {
         if (line[0] == 'T') {
             int result = sscanf(line, "%*c %79c", title);
